AEM prerender docs for developers and merchants (#543)
commerce-docs/microsite-commerce-storefront: src/content/docs/setup/configuration/aem-prerender.mdx

---
title: AEM Commerce Prerender
description: Learn how to implement server-side rendering for product pages using the AEM Commerce Prerender App Builder app for improved SEO and LLM crawlability.
---

import Diagram from '@components/Diagram.astro';
import Link from '@components/Link.astro';
import TableWrapper from '@components/TableWrapper.astro';
import { Steps } from '@astrojs/starlight/components';
import { Aside } from '@astrojs/starlight/components';

The AEM Commerce Prerender solution enables your Edge Delivery Services Commerce Storefront to serve fully-rendered product pages to search engines and AI crawlers before JavaScript executes. This improves SEO rankings and makes your product catalog accessible to LLMs.

> **How it works:** You deploy an App Builder application that runs continuously in Adobe's cloud, monitoring your catalog and generating complete HTML for each product. This HTML publishes to Edge Delivery Services, where your storefront serves it to visitors. Your storefront's product-details block then enhances the prerendered HTML with interactive features powered by the PDP drop-in. You must deploy the App Builder app first because your storefront integration depends on the HTML it generates.

Reviewer: ```suggestion
**How it works:** You deploy an App Builder application that runs continuously in Adobe's cloud, monitoring your catalog and generating complete HTML for each product. This HTML publishes to Edge Delivery Services, where your storefront serves it to visitors. Your storefront `product-details` block then enhances the prerendered HTML with interactive features powered by the PDP drop-in. You must deploy the App Builder app first because your storefront integration depends on the HTML it generates.
```

- **Improved Performance**: Reduces the time-to-first-byte and serves immediate content
- **Natural Fallback**: Pre-rendered content displays while client-side drop-ins load
- **Automated Updates**: Monitors the product catalog and updates markup when products change

## Implementation roadmap

This guide walks you through implementing the prerender solution in both required repositories. Follow these steps in order:

<Diagram 
  type="mermaid" 
  caption="Complete implementation workflow showing the step-by-step process for setting up prerendered product pages in your storefront."
  code={`%%{init: {'theme':'base', 'themeVariables': {'fontSize':'16px'}}}%%
flowchart LR
    Start([<strong>Start Implementation</strong>]):::start
    
    Step1["<strong>1. Check Prerequisites</strong><br/>Verify access, tools, and storefront requirements"]:::step
    Step2["<strong>2. Understand the Architecture</strong><br/>Learn how the two repositories work together"]:::step
    Step3["<strong>3. Install & Deploy App Builder App</strong><br/>Repository: aem-commerce-prerender<br/>Clone, configure, and deploy to App Builder"]:::repo1

>     Step4["<strong>4. Integrate with Your Storefront</strong><br/>Repository: Your Commerce Boilerplate<br/>Modify product-details block to use prerendered markup"]:::repo2

Reviewer: ```suggestion
    Step4["<strong>4. Integrate with Your Storefront</strong><br/>Repository: Your Commerce Boilerplate<br/>Modify the `product-details` block to use prerendered markup"]:::repo2
```

    Step5["<strong>5. Customize (Optional)</strong><br/>Adjust templates, schemas, and queries"]:::optional
    
    End([<strong>✓ Implementation Complete</strong>]):::complete
    
    Start --> Step1
    Step1 --> Step2
    Step2 --> Step3
    Step3 --> Step4
    Step4 --> Step5
    Step5 --> End
    
    classDef start fill:#e3f2fd,stroke:#1976d2,stroke-width:3px,color:#000
    classDef step fill:#f5f5f5,stroke:#666,stroke-width:2px,color:#000
    classDef repo1 fill:#fff3e0,stroke:#f57c00,stroke-width:3px,color:#000
    classDef repo2 fill:#e8f5e9,stroke:#388e3c,stroke-width:3px,color:#000
    classDef optional fill:#fce4ec,stroke:#c2185b,stroke-width:2px,color:#000,stroke-dasharray: 5 5
    classDef complete fill:#c8e6c9,stroke:#2e7d32,stroke-width:3px,color:#000
    
    linkStyle default stroke-width:3px,stroke:#666
`}
/>

## Prerequisites

Before implementing the prerender solution, verify you have:

> ### Required Access and Permissions

Reviewer: Use sentence case for all headings. Currently, some are initial caps; some are sentence case

```suggestion
### Required access and permissions
```

- **Adobe Developer Console access** with "Developer" role for your organization
- **App Builder workspace** (Stage and Production workspaces recommended)

> - **AEM Admin API access** for your Edge Delivery Services project

Reviewer: Consider including a reference to the AEM Admin API documentation: https://www.aem.live/docs/admin.html#tag/status/operation/status

- **Adobe Commerce instance** with Catalog Service configured and operational

### Storefront Requirements

Your Edge Delivery Services Commerce Storefront must be:

- Built on the <Link href="https://github.com/hlxsites/aem-boilerplate-commerce" text="Commerce boilerplate" />
- Configured with valid [storefront configuration](/setup/configuration/commerce-configuration/) connecting to your Adobe Commerce backend
- Using product URLs with consistent format (typically `/products/{urlKey}/{sku}`)

### Technical Requirements

- **Node.js** (current LTS version) installed locally
- **Adobe I/O CLI** (`@adobe/aio-cli`) installed globally
- **Git** for cloning the prerender repository
- Familiarity with JavaScript, Handlebars templates, and GraphQL

## How it works

The prerender system operates as a continuous background process:

<Steps>

1. The store admin manages products in Adobe Commerce
2. The Prerender Stack detects product changes
3. The Prerender Stack renders pages with semantic markup
4. The pages are published via the AEM Admin API
5. The sitemap is updated via the AEM Admin API

> 6. Store customers access the pre-rendered product details pages

Reviewer: ```suggestion
1. The store admin manages products in Adobe Commerce.
2. The Prerender Stack detects product changes.
3. The Prerender Stack renders pages with semantic markup.
4. The pages are published via the AEM Admin API.
5. The sitemap is updated via the AEM Admin API.
6. Store customers access the pre-rendered product details pages.
```

- **Catalog Service Integration**: Queries product data from your Adobe Commerce backend
- **AEM Admin API**: Publishes generated markup to your Edge Delivery Services project
- **Storage Layer**: Maintains product state and generated markup in App Builder storage
- **Management UI**: Web interface for monitoring, configuration, and troubleshooting

<Diagram 
  type="mermaid" 
  caption="Complete architecture showing App Builder actions, content management, and how pre-rendered and dynamic content are delivered to consumers."
  code={`flowchart TB
    subgraph subGraph0["<strong>App Builder Workspace</strong>"]
        direction TB
        A["Product Change Detector<br/>(publishes products upon change)"]:::appBuilder
        B["Product List Scraper<br/>(fetch and store list)"]:::appBuilder
        C["PDP Renderer<br/>(generates/stores pages)"]:::appBuilder
        D[("Blob Store")]:::storage
        A --> D
        B --> D
        C --> D
    end

    E["public/ folder<br/>PDPs served publicly"]:::public
    F["Edge Delivery Services<br/>fetches HTML from public blob store folder"]:::helix
    
    subgraph subGraph1["<strong>Product Page</strong>"]
        direction TB

>         G["Embedded markup with product data<br/>and attributes, injected ahead-of-time"]:::embedded

Reviewer: ```suggestion
        G["Embedded markup with product data<br/>and attributes, injected ahead of time"]:::embedded
```

`}
/>

<Aside type="note" title="Catalog Size Considerations">
For catalogs exceeding 50,000 products, consult the <Link href="https://www.aem.live/docs/limits" text="AEM indexing limits" /> documentation for sitemap splitting strategies and scaling considerations.
</Aside>

## Installation and configuration

The <Link href="https://github.com/adobe-rnd/aem-commerce-prerender" text="AEM Commerce Prerender repository" /> provides complete installation instructions. The following overview highlights the key configuration steps and values you'll need.

<Steps>

1. **Clone the repository** and install dependencies:
   ```bash
   git clone https://github.com/adobe-rnd/aem-commerce-prerender.git
   cd aem-commerce-prerender
   npm install
   ```

2. **Run the setup wizard** to generate your `.env` configuration:
   ```bash
   npm run setup
   ```
   The wizard prompts for your App Builder workspace selection and Adobe I/O authentication.

3. **Configure environment variables** in the generated `.env` file with your storefront-specific values:

   <TableWrapper nowrap={[0]}>

   | Variable | Description | Example |
   |----------|-------------|---------|

>    | `CATALOG_SERVICE_URL` | Your Adobe Commerce Catalog Service endpoint | `https://catalog-service.adobe.io/graphql` |

Reviewer: For Adobe Commerce Optimizer, this URL is different. See  endpoints:  https://experienceleague.adobe.com/developer/commerce/storefront/setup/configuration/commerce-configuration/#endpoints

   | `AEM_ADMIN_API_KEY` | Admin API token from your <Link href="https://www.aem.live/docs/admin.html#setting-up-authentication-for-admin-api" text="AEM project settings" /> | `your-admin-token-here` |
   | `AEM_ORG` | Your AEM organization identifier (e.g., `adobe`) | `adobe` |
   | `AEM_SITE` | Your AEM site name (e.g., `your-storefront`) | `your-storefront` |
   | `PRODUCT_PAGE_URL_FORMAT` | URL pattern for product pages | `/products/{urlKey}/{sku}` |
   | `COMMERCE_ENVIRONMENT_ID` | Commerce instance environment ID from Commerce Admin | `xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx` |
   | `COMMERCE_WEBSITE_CODE` | Website code from Commerce multistore setup | `base` or your store code |

>    | `COMMERCE_STORE_VIEW_CODE` | Store view code from Commerce | `default` or your view code |

Reviewer: If Commerce pre-render works with an ACO backend, then these headers would be different I believe.  See [Storefront configuration](https://experienceleague.adobe.com/developer/commerce/storefront/setup/configuration/commerce-configuration/).

   </TableWrapper>

4. **Obtain required credentials**:
   - **AEM Admin API token**: Generate from your <Link href="https://www.aem.live/docs/admin.html#setting-up-authentication-for-admin-api" text="AEM project admin panel" />

>    - **Commerce credentials**: Access your Commerce Admin → System → Services → <Link href="https://experienceleague.adobe.com/en/docs/commerce-merchant-services/user-guides/integration-services/saas" text="Commerce Services Connector" /> for environment and API configuration

Reviewer: If a  customer is using ACO with Adobe Commerce on cloud or on-premises via the ACO connector, you would configure the ACO connector to get these credentials.

I think we could publish this for the Adobe Commerce as a Cloud Service context, and then update it to add the ACO context.

The Management UI provides:

- **Monitoring capabilities** to view published products and track the product index
- **Change detector controls** to start, stop, and configure scheduled product polling
- **Markup preview** to view generated HTML for product pages before publication
- **Logs and activations** for debugging App Builder action executions
- **Storage management** to trigger manual product operations and manage blob storage
- **Configuration settings** to review environment variables and deployment parameters

Access and authentication details are configured during the initial deployment process. Refer to the <Link href="https://github.com/adobe-rnd/aem-commerce-prerender" text="repository documentation" /> for setup instructions.

## Storefront integration

After installing the prerender solution, modify your Commerce boilerplate to integrate prerendered markup with your product detail pages.

### Understanding the Two-Repository Architecture

The prerender solution involves two separate codebases that work together:

1. **The aem-commerce-prerender repository** (App Builder app)

>    - This is what you cloned, configured, and deployed in the previous section

Reviewer: ```suggestion
   - This repository is the one that you cloned, configured, and deployed in the previous section
```

   - Runs as a serverless application in Adobe App Builder
   - Generates prerendered HTML and publishes it to your Edge Delivery Services site
   - Operates independently as a background process

2. **Your storefront repository** (Commerce boilerplate)
   - Your customer-facing website code
   - Built on the <Link href="https://github.com/hlxsites/aem-boilerplate-commerce" text="Commerce boilerplate" />
   - Needs code modifications to properly consume the prerendered HTML

>    - This is where you'll make the integration changes described below

Reviewer: ```suggestion
   - This repository is where you make the integration changes described below
```

> **How they connect:** The App Builder app publishes prerendered HTML to Edge Delivery Services at URLs matching your product pages (e.g., `/products/acme-widget/SKU123`). When a visitor (or crawler) requests a product page, Edge Delivery Services serves this prerendered HTML. Your storefront's JavaScript then enhances the page by replacing the static HTML with the interactive PDP drop-in.

Reviewer: ```suggestion
**How they connect:** The App Builder app publishes prerendered HTML to Edge Delivery Services at URLs matching your product pages (for example, `/products/acme-widget/SKU123`). When a visitor (or crawler) requests a product page, Edge Delivery Services serves this prerendered HTML. Your storefront JavaScript then enhances the page by replacing the static HTML with the interactive PDP drop-in.
```

  type="mermaid" 
  caption="Two-repository architecture showing how the App Builder prerender app publishes HTML that your storefront code consumes and enhances for visitors."
  code={`%%{init: {'theme':'base', 'themeVariables': {'lineColor':'#333','edgeLabelBackground':'#ffffff'}, 'flowchart':{'curve':'basis','defaultRenderer':'dagre','rankSpacing':50,'nodeSpacing':50}}}%%
flowchart LR
    subgraph repos["<strong>Developer Repositories</strong>"]
        direction TB
        subgraph repo1[" "]
            direction TB
            A0["<strong>Repository 2</strong><br/>aem-commerce-prerender<br/>(App Builder Application)"]:::title
            A1["<span style='font-size:60px'>👨‍💻</span><br/><span style='font-size:18px'>Developer Actions</span>"]:::icon
            A2["Clone repo<br/>Configure .env<br/>Deploy to App Builder"]:::action
            A0 -.-> A1
            A1 --> A2
        end
        
        subgraph repo2[" "]
            direction TB
            D0["<strong>Repository 1</strong><br/>Your Storefront<br/>(Commerce Boilerplate)"]:::title
            D1["<span style='font-size:60px'>👨‍💻</span><br/><span style='font-size:18px'>Developer Actions</span>"]:::icon

>             D2["Modify product-details block<br/>to read prerendered HTML<br/>and enhance with PDP drop-in"]:::action

Reviewer: Use `product-details` block  consistently. It's easier to identify it as a code component.  Current usage is inconsistent.

    classDef action fill:#fff3e0,stroke:#f57c00,stroke-width:2px
    classDef process fill:#f3e5f5,stroke:#7b1fa2,stroke-width:2px
    classDef storage fill:#e8f5e9,stroke:#388e3c,stroke-width:2px
    classDef delivery fill:#fff9c4,stroke:#f57f17,stroke-width:2px
    classDef crawlerEnd fill:#c8e6c9,stroke:#388e3c,stroke-width:2px
    
    style repos fill:#e3f2fd,stroke:#1976d2,stroke-width:2px
    style repo1 fill:#f3e5f5,stroke:#7b1fa2,stroke-width:1px
    style repo2 fill:#fff3e0,stroke:#f57c00,stroke-width:1px
    style appbuilder fill:#e8f5e9,stroke:#388e3c,stroke-width:2px
    style eds fill:#e3f2fd,stroke:#1976d2,stroke-width:2px
    style delivery fill:#fce4ec,stroke:#c2185b,stroke-width:2px
    
    linkStyle default stroke-width:3px,stroke-dasharray:5
`}
/>

### How Prerendered HTML Reaches Your Storefront

The prerendered HTML flows from the App Builder app to your storefront through the AEM Admin API:

**Publishing mechanism:**

<Steps>

1. **The App Builder app generates HTML** → Creates complete HTML files with prerendered product markup
2. **API call to AEM Admin** → The App Builder app makes authenticated API calls using your `AEM_ADMIN_API_KEY`

> 3. **Content published to EDS** → The Admin API publishes the HTML files to Edge Delivery Services at the specified product URLs

Reviewer: Make sure that references to the Admin API include AEM Admin API to avoid any confusion with Commerce APIs

```suggestion
3. **Content published to EDS** → The AEM Admin API publishes the HTML files to Edge Delivery Services at the specified product URLs
```

> 4. **Sitemap updated** → The Admin API also updates your sitemap to include the new/updated product pages

Reviewer: ```suggestion
4. **Sitemap updated** → The AEM Admin API also updates your sitemap to include the new and updated product pages
```

5. **Content available** → Edge Delivery Services now has the HTML files stored and ready to serve

</Steps>

> This is **not an event-based system**: it's a direct API integration. Each time the App Builder app detects a product change and generates HTML, it immediately publishes that HTML to Edge Delivery Services via API calls. Your storefront doesn't need to do anything special to "receive" these files; they're automatically available at the product URLs once published.

Reviewer: ```suggestion
The Commerce Prerender app is a direct API integration. It is **not an event-based system**. Each time the App Builder app detects a product change and generates HTML, it immediately publishes that HTML to Edge Delivery Services via API calls. When files are published, they are exposed at the product URLs automatically; the storefront does not need to perform any additional fetching or configuration.
```

### How the Storefront Serves Prerendered HTML

> When a visitor requests a product page URL (e.g., `/products/acme-widget/sku123`), Edge Delivery Services serves the prerendered HTML as the actual page. The HTML already contains the `product-details` block populated with prerendered content; your storefront doesn't fetch it separately.

Reviewer: ```suggestion
When a visitor requests a product page URL (for example, `/products/acme-widget/sku123`), Edge Delivery Services serves the prerendered HTML as the actual page. The HTML already contains the `product-details` block populated with prerendered content; your storefront doesn't fetch it separately.
```

**The delivery sequence:**

<Steps>

1. **Request arrives** → Edge Delivery Services receives the URL request
2. **HTML served** → EDS returns the prerendered HTML file (published by the App Builder app)
3. **Page loads** → The browser displays the page with prerendered content (crawlers stop here)
4. **JavaScript executes** → Your storefront code runs
5. **Enhancement** → The `product-details` block's `decorate()` function replaces static HTML with interactive UI

</Steps>

> This is **progressive enhancement**: the prerendered HTML is the initial page that everyone sees. Your storefront code simply finds what's already in the DOM, extracts the SKU from the meta tag, and upgrades it to an interactive experience for users with JavaScript. This approach ensures crawlers always see complete product information, users with JavaScript enabled get the full interactive experience, and users without JavaScript still see product details (though without dynamic features).

Reviewer: ```suggestion
This process is called **progressive enhancement**: The prerendered HTML is the initial page that everyone sees. The storefront code reads the existing DOM, extracts the SKU from the meta tag, and upgrades the page to an interactive experience for users with JavaScript. This approach ensures that crawlers and users without Javascript always see complete product information, while users with JavaScript get the full interactive experience.
```

### Modifying Your Product Details Block

Developers must modify the product-details block in their <Link href="https://github.com/hlxsites/aem-boilerplate-commerce" text="Commerce boilerplate" /> to integrate prerendered markup with the PDP drop-in.

#### Location

The <Link href="https://github.com/hlxsites/aem-boilerplate-commerce/tree/main/blocks/product-details" text="product-details block" /> lives in your boilerplate at:

```
blocks/product-details/product-details.js
```

#### Integration Pattern

> Your block's `decorate()` function should:

Reviewer: ```suggestion
Your block `decorate()` function should:
```

<Steps>

1. **Preserve prerendered markup**: The block already contains server-rendered HTML visible to crawlers before JavaScript executes.
2. **Retrieve the SKU**: Get the product SKU from the `<meta name="sku">` tag instead of parsing the sanitized URL.
3. **Replace with PDP drop-in**: Clear the prerendered content and render the interactive PDP component.

</Steps>

```js
export default async function decorate(block) {
  // Step 1: Preserve prerendered markup
  // The block already contains HTML. No action needed here;
  // crawlers see this content before JavaScript runs
  
  // Step 2: Retrieve the SKU from meta tag
  const skuMeta = document.querySelector('meta[name="sku"]');
  const sku = skuMeta?.content || getSkuFromUrl();
  
  // Step 3: Replace with PDP drop-in
  block.innerHTML = ''; // Clear prerendered content
  await productRenderer.render(ProductDetails, { sku })(block); // Render interactive UI
}
```

#### URL Format and SKU Handling

> The prerender system generates markup files that match your storefront's product URL structure. By default, this uses the format:

Reviewer: ```suggestion
The prerender system generates markup files that match your storefront product URL structure. By default, this uses the format:
```

|---------------|---------------------------|
| **Templates** | Control HTML markup structure and styling by modifying `tools/templates/pdp.handlebars` |
| **Structured Data** | Customize JSON-LD schemas in `tools/structured-data/product.js` to optimize product rich snippets for search engines |
| **GraphQL Queries** | Extend product data by adjusting Catalog Service queries in `actions/pdp-renderer/index.js` to fetch additional attributes |
| **URL Patterns** | Match your storefront URL structure by configuring `PRODUCT_PAGE_URL_FORMAT` in the `.env` file |
| **Rendering Logic** | Transform product data before markup generation by implementing custom logic in App Builder actions |

</TableWrapper>

For implementation details, see the <Link href="https://github.com/adobe-rnd/aem-commerce-prerender" text="AEM Commerce Prerender repository" />, particularly the `actions/`, `tools/`, and repository root for templates and configuration files.

## Troubleshooting

For troubleshooting during development and operations, consult the <Link href="https://github.com/adobe-rnd/aem-commerce-prerender/blob/main/docs/RUNBOOK.md" text="prerender runbook" />.

Common issues include:

> - **Authentication failures**: Verify your AEM Admin API token is valid (tokens expire after 1 year)

Reviewer: ```suggestion
- **Authentication failures**: Verify that your AEM Admin API token is valid (tokens expire after 1 year)
```

  - Check token expiration in your <Link href="https://www.aem.live/docs/admin.html#setting-up-authentication-for-admin-api" text="AEM project admin panel" />
  - Generate a new token if expired and update `AEM_ADMIN_API_KEY` in your `.env` file
  - Redeploy with `npm run deploy` after updating credentials
  

> - **Product not rendering**: Check that the product exists in Catalog Service and the SKU matches

Reviewer: ```suggestion
- **Product not rendering**: Check that the product exists in Catalog Service and that the SKU matches
```

>   - Query Catalog Service directly using its <Link href="https://developer.adobe.com/commerce/services/graphql/catalog-service/products/" text="GraphQL Product API" /> to verify product data

Reviewer: ```suggestion
  - Query the Catalog Service directly using its <Link href="https://developer.adobe.com/commerce/services/graphql/catalog-service/products/" text="GraphQL Product API" /> to verify product data
```

>   - Use the Management UI to view the product index and confirm the SKU is listed

Reviewer: ```suggestion
  - Use the Management UI to view the product index and confirm that the SKU is listed
```

  
- **Markup not updating**: Review the Change Detector logs to ensure scheduled triggers are running
  - Access the <Link href="https://prerender.aem-storefront.com" text="Management UI" /> to verify the change detector status (should show "Running")
  - Check activation logs: `aio rt activation list product-change-detector`
  - View detailed logs: `aio rt activation logs <ACTIVATION_ID>`
  - Manually trigger the change detector from the Management UI to test functionality
  
- **Storage errors**: Verify your App Builder workspace has sufficient storage quota
  - Check storage usage in the Adobe Developer Console under your App Builder workspace
  - Review blob storage logs: `aio rt activation list blob-store`
  - Contact Adobe support to request quota increases if needed

## Additional resources

- **App Builder Documentation**: <Link href="https://developer.adobe.com/app-builder/docs/overview/" text="Adobe Developer App Builder" />
- **Catalog Service Guide**: <Link href="https://experienceleague.adobe.com/en/docs/commerce/catalog-service/installation" text="Adobe Commerce Catalog Service" />

> - **Edge Delivery Services**: <Link href="https://www.aem.live/docs/publishing-from-authoring" text="Publishing to Edge Delivery Services" />

Reviewer: Do you need to include the AEM Admin API documentation here -- or maybe it's not required because those calls are made automatically once configured?

> ---

Reviewer 2: The github repo for prerender includes a link to [this page](https://www.aem.live/developer/byom) but it might be good to also link it here, as a "pre-read" topic.

title: AEM Commerce Prerender
description: Learn how to implement server-side rendering for product pages using the AEM Commerce Prerender App Builder app for improved SEO and LLM crawlability.
---

import Diagram from '@components/Diagram.astro';
import Link from '@components/Link.astro';
import LinkCard from '@components/LinkCard.astro';
import TableWrapper from '@components/TableWrapper.astro';
import { Steps, Aside } from '@astrojs/starlight/components';

> The AEM Commerce Prerender solution enables your Edge Delivery Services Commerce Storefront to serve fully-rendered product pages to search engines and AI crawlers before JavaScript executes. This improves SEO rankings and makes your product catalog accessible to LLMs.

Reviewer 2: ```suggestion
The AEM Commerce Prerender solution enables your Edge Delivery Services Commerce Storefront to serve data-hydrated product page markup to search engines and AI crawlers _before_ JavaScript executes. This can improve SEO rankings and makes your product catalog accessible to LLMs.
```

> **How it works:** You deploy an App Builder application that runs continuously in Adobe's cloud, monitoring your catalog and generating complete HTML for each product. This HTML publishes to Edge Delivery Services, where your storefront serves it to visitors. Your storefront `product-details` block then enhances the prerendered HTML with interactive features powered by the PDP drop-in. You must deploy the App Builder app first because your storefront integration depends on the HTML it generates.

Reviewer 2: ```suggestion
**How it works:** You deploy an App Builder application that runs continuously in Adobe's cloud, monitoring your catalog and generating complete HTML for each product. This HTML is published to Edge Delivery Services and served as the initial HTML "source" to page visitors. Then, your `product-details` block enhances the pre-rendered HTML with interactive features powered by the PDP drop-in. You should the App Builder app first because your storefront integration depends on the HTML it generates.
```

I think we should add some detail, or to this page somewhere, about the "fallback"  content flow with EDS. I don't know it but @buuhuu might.

Essentially if you have overlay configured for say `/products/fooSku`, then a request to the page will first try to get the source from the overlay for the path, and fall back to looking for the actual document at the same path within the content source (da.live).

## Why use prerender?

> Modern storefronts use client-side JavaScript to render product information, limiting discoverability by search engines and AI systems that don't execute JavaScript. Prerendering solves this by generating complete HTML with structured data before pages are requested.

Reviewer 2: "Modern storefronts use..." not true. More like "Adobe Commerce on EDS historically used client side JS to render product information".

The truth is, we (Adobe/AEM) have to introduce prerender since there is no server rendering. We should probably make it quite clear that EDS has no concept of "server rendering" in it's stack - the closest thing is this, BYOM.

`}
/>

<Aside type="note" title="Catalog Size Considerations">
For catalogs exceeding 50,000 products, consult the <Link href="https://www.aem.live/docs/limits" text="AEM indexing limits" /> documentation for sitemap splitting strategies and scaling considerations.
</Aside>

## Installation and configuration

The <Link href="https://github.com/adobe-rnd/aem-commerce-prerender" text="AEM Commerce Prerender repository" /> provides complete installation instructions. The following overview highlights the key configuration steps and values you'll need.

<Steps>

1. **Clone the repository** and install dependencies:
   ```bash
   git clone https://github.com/adobe-rnd/aem-commerce-prerender.git
   cd aem-commerce-prerender
   npm install
   ```

2. **Run the setup wizard** to generate your `.env` configuration:
   ```bash
   npm run setup
   ```
   The wizard prompts for your App Builder workspace selection and Adobe I/O authentication.

3. **Configure environment variables** in the generated `.env` file with your storefront-specific values:

   <TableWrapper nowrap={[0]}>

   | Variable | Description | Example |
   |----------|-------------|---------|

>    | `CATALOG_SERVICE_URL` | Your Adobe Commerce Catalog Service endpoint | `https://catalog-service.adobe.io/graphql` |

Reviewer 3: This variable doesn’t need to be configured in the .env file — it’s usually defined in the configuration file

   | `AEM_ADMIN_API_KEY` | Admin API token from your <Link href="https://www.aem.live/docs/admin.html#setting-up-authentication-for-admin-api" text="AEM project settings" /> | `your-admin-token-here` |
   | `AEM_ORG` | Your AEM organization identifier (e.g., `adobe`) | `adobe` |
   | `AEM_SITE` | Your AEM site name (e.g., `your-storefront`) | `your-storefront` |
   | `PRODUCT_PAGE_URL_FORMAT` | URL pattern for product pages | `/products/{urlKey}/{sku}` |
   | `COMMERCE_ENVIRONMENT_ID` | Commerce instance environment ID from Commerce Admin | `xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx` |
   | `COMMERCE_WEBSITE_CODE` | Website code from Commerce multistore setup | `base` or your store code |

>    | `COMMERCE_STORE_VIEW_CODE` | Store view code from Commerce | `default` or your view code |

Reviewer 3: These 3 params also should not be in .env , please check what has to be in the .env file here: https://github.com/adobe-rnd/aem-commerce-prerender. Also, please check variable names

   </TableWrapper>

4. **Obtain required credentials**:
   - **AEM Admin API token**: Generate from your <Link href="https://www.aem.live/docs/admin.html#setting-up-authentication-for-admin-api" text="AEM project admin panel" />
   - **Commerce credentials**: Access your Commerce Admin → System → Services → <Link href="https://experienceleague.adobe.com/en/docs/commerce-merchant-services/user-guides/integration-services/saas" text="Commerce Services Connector" /> for environment and API configuration

>    - **App Builder credentials**: Automatically configured when you authenticate via the setup wizard

Reviewer 3: For these credentials to take effect, you need to download the configuration file from AppBuilder and run aio app use file_name in the root directory of your project.

5. **Deploy to App Builder**:
   ```bash
   npm run deploy
   ```
   This deploys all actions and schedules to your selected workspace (Stage or Production).

6. **Test the deployment**:

>    - Manually invoke actions using the Management UI or `aio rt action invoke`

Reviewer 3: it can also be done via CLI

   - Verify product markup generation for sample SKUs
   - Check that pages publish successfully to your AEM project

7. **Enable automated synchronization**:
   - Use the <Link href="https://prerender.aem-storefront.com" text="Management UI" /> to start the change detector

>    - Configure polling intervals (default: checks for changes every 5 minutes)

Reviewer 3: Maybe we should mention that this needs to be specified in the app.config.yaml file.

## Storefront integration

After installing the prerender solution, modify your Commerce boilerplate to integrate prerendered markup with your product detail pages.

### Understanding the Two-Repository Architecture

The prerender solution involves two separate codebases that work together:

1. **The aem-commerce-prerender repository** (App Builder app)
   - This repository is the one that you cloned, configured, and deployed in the previous section
   - Runs as a serverless application in Adobe App Builder
   - Generates prerendered HTML and publishes it to your Edge Delivery Services site
   - Operates independently as a background process

2. **Your storefront repository** (Commerce boilerplate)
   - Your customer-facing website code
   - Built on the <Link href="https://github.com/hlxsites/aem-boilerplate-commerce" text="Commerce boilerplate" />
   - Needs code modifications to properly consume the prerendered HTML
   - This repository is where you make the integration changes described below

> **How they connect:** The App Builder app publishes prerendered HTML to Edge Delivery Services at URLs matching your product pages (for example, `/products/acme-widget/SKU123`). When a visitor (or crawler) requests a product page, Edge Delivery Services serves this prerendered HTML. Your storefront JavaScript then enhances the page by replacing the static HTML with the interactive PDP drop-in.

Reviewer 3: App Builder publishes the HTML to its own bucket, and in EDS, the link to that bucket is specified in the configuration file as an overlay also, in the url  the SKU should be written in lowercase because uppercase letters are not supported.

#### Location

The <Link href="https://github.com/hlxsites/aem-boilerplate-commerce/tree/main/blocks/product-details" text="product-details block" /> lives in your boilerplate at:

```
blocks/product-details/product-details.js
```

#### Integration Pattern

Your block `decorate()` function should:

<Steps>

1. **Preserve prerendered markup**: The block already contains server-rendered HTML visible to crawlers before JavaScript executes.
2. **Retrieve the SKU**: Get the product SKU from the `<meta name="sku">` tag instead of parsing the sanitized URL.
3. **Replace with PDP drop-in**: Clear the prerendered content and render the interactive PDP component.

</Steps>

```js
export default async function decorate(block) {
  // Step 1: Preserve prerendered markup
  // The block already contains HTML. No action needed here;
  // crawlers see this content before JavaScript runs
  
  // Step 2: Retrieve the SKU from meta tag
  const skuMeta = document.querySelector('meta[name="sku"]');

>   const sku = skuMeta?.content || getSkuFromUrl();

Reviewer 2: We have a [getProductSku function](https://github.com/hlxsites/aem-boilerplate-commerce/blob/5e79196c4d792fca8b8d689c91e52a0649062669/scripts/commerce.js#L607) which already does this in commerce boilerplate.

I leave it to you whether you want to reference the provided function or be a bit more verbose for clarity.

Author: Added

<Aside type="note" title="SKU Lowercase Requirement">
Starting with the October 2025 Adobe Commerce Storefront release, all SKUs in product URLs are automatically converted to lowercase to ensure URL consistency and proper resolution. When the prerendered markup is generated, the actual SKU (with original casing) is stored in a meta tag:

```html
<meta name="sku" content="MY_PRODUCT_123">
```

This allows your PDP drop-in to retrieve the correct SKU regardless of URL transformations, which is why the code example above retrieves the SKU from the meta tag rather than parsing it from the URL.
</Aside>

For complete implementation examples, refer to the <Link href="https://github.com/hlxsites/aem-boilerplate-commerce/tree/main/blocks/product-details" text="product-details block" /> in the Commerce boilerplate.

## Customization

The prerender solution provides these customization options:

<TableWrapper nowrap={[0]}>

| Customization | Purpose and Configuration |
|---------------|---------------------------|

> | **Templates** | Control HTML markup structure and styling by modifying `tools/templates/pdp.handlebars` |

Reviewer 3: The markup is handled on the client side at store_url/product/default by default. If it’s not available, the template can be edited here: [product-details.hbs](https://github.com/adobe-rnd/aem-commerce-prerender/blob/main/actions/pdp-renderer/templates/product-details.hbs)
.

> | **Structured Data** | Customize JSON-LD schemas in `tools/structured-data/product.js` to optimize product rich snippets for search engines |

Reviewer 3: it is not correct , idjson may be modified here: https://github.com/adobe-rnd/aem-commerce-prerender/blob/main/actions/pdp-renderer/ldJson.js

> | **GraphQL Queries** | Extend product data by adjusting Catalog Service queries in `actions/pdp-renderer/index.js` to fetch additional attributes |

Reviewer 3: The logic is implemented in the generateProductHtml method located in [render.js](https://github.com/adobe-rnd/aem-commerce-prerender/blob/main/actions/pdp-renderer/render.js)
, while the actual GraphQL queries are defined in [queries.js](https://github.com/adobe-rnd/aem-commerce-prerender/blob/main/actions/queries.js)
.

| **URL Patterns** | Match your storefront URL structure by configuring `PRODUCT_PAGE_URL_FORMAT` in the `.env` file |
| **Rendering Logic** | Transform product data before markup generation by implementing custom logic in App Builder actions |

</TableWrapper>

> For implementation details, see the <Link href="https://github.com/adobe-rnd/aem-commerce-prerender" text="AEM Commerce Prerender repository" />, particularly the `actions/`, `tools/`, and repository root for templates and configuration files.

Reviewer 3: actions is the main folder, please remove tools

## Troubleshooting

For troubleshooting during development and operations, consult the <Link href="https://github.com/adobe-rnd/aem-commerce-prerender/blob/main/docs/RUNBOOK.md" text="prerender runbook" />.

Common issues include:

- **Authentication failures**: Verify that your AEM Admin API token is valid (tokens expire after 1 year)

>   - Check token expiration in your <Link href="https://www.aem.live/docs/admin.html#setting-up-authentication-for-admin-api" text="AEM project admin panel" />

Reviewer 3: To verify the token, you can use https://www.jwt.io/ — simply decode the token and check its expiration time.

>   - Generate a new token if expired and update `AEM_ADMIN_API_KEY` in your `.env` file

Reviewer 3: here is the link how to generate a new token: https://www.aem.live/docs/admin.html#tag/orgConfig/operation/createOrgApiKey
